### DISCUSSION TITLE
No nonce on script element in body added by Next chunk preloading?

### DISCUSSION THREAD

**Discussion by ludwighogstrom** [vercel/next.js#28658] [Help]
Hi! 

Trying to implement `strict dynamic` CSP rules, suggested [here](https://web.dev/strict-csp/), to *avoid* "allow listening" of domains.

Seems to work in Chrome so far but I have stumble across a problem in Safari. Safari doesn't support CSP level 3/`strict dynamic`, instead it uses a fallback rule based on `nonce`. 

Error is a bit vague but I think I have problem because Next preload chunks via script elements added to the bottom of the body (when using the link component). Those script elements does not have a `nonce` attribute.

Is there any way to add `nonce` to those scripts?

<img width="592" alt="Skärmavbild 2021-08-31 kl  15 21 04" src="https://user-images.githubusercontent.com/3797543/131511075-8eedd909-e6b8-4904-bbfd-9aabfa7a03a2.png">

CSP error in Safari:
```
Refused to execute a script because its hash, its nonce, or 'unsafe-inline' does not appear in the script-src directive of the Content Security Policy.
```

CSP error report:
```
{
   "csp-report":{
      "document-uri":"http://localhost:3000/sv-se",
      "referrer":"",
      "violated-directive":"script-src 'nonce-foo-123456789' 'strict-dynamic' 'unsafe-inline' 'unsafe-eval' 'self' https: http:",
      "effective-directive":"script-src",
      "original-policy":"object-src 'none'; report-to csp-endpoint; report-uri /api/report-csp-violation; base-uri 'none'; script-src 'nonce-foo-123456789' 'strict-dynamic' 'unsafe-inline' 'unsafe-eval' 'self' https: http:;",
      "blocked-uri":"",
      "status-code":200
   }
}
```

**Comment by leerob** [comment]
Hey folks, wanted to swing back here with an update. After digging through many different issues and discussions, I've [**made a new page in the documentation**](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy) ([PR](https://github.com/vercel/next.js/pull/54601)) specifically for Content Security Policy and nonces. This docs page: 

- Explains how to generate a `nonce` with Middleware
- Shows how to consume the `nonce` in a route with `headers()`
- Shows a complete CSP without needing to use any `unsafe`
- Shows how to ignore the `nonce` Middleware from running on prefetches / static assets

Further, we've patched some bugs and made improvements to `nonce` handling in Next.js that will be available in the latest `canary` version (for those of you time traveling from the future, upgrade to Next.js 13.5). We also updated the `with-strict-csp` example in the `examples/` folder, which is backlinked from the [new documentation page](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy).

Really hope this helps out, thank you all 🙏 I'll be closing this discussion out. To continue the discussion, [please go here](https://github.com/vercel/next.js/discussions/54907).
